### PR TITLE
Display rejected games in a table

### LINKED ISSUE #1747: Creator gets no indication of a rejected game
If I create a game, and you reject it, I don't get any notification or other indication that this has happened. This breaks a really common use case for me: You and I finish game A, I create a new game B (e.g. with random buttons), you reject it because you don't like the buttons it picked. What I'd like to be able to do is click on something to try again; that could just be a link to the game B page, from which I could go back to the previous game, and then click "new game with random buttons" again, or maybe something more helpful with less clicking. But having it just silently vanish is terrible, because it means the chain of games we've been playing is broken, until I notice that we're not playing any more, and go try again.

### LINKED ISSUE #1758: Missing verb on preferences page: "These the colors used to represent each player in a game."
The text "These the colors used to represent each player in a game." probably should a verb near the beginning. Perhaps "are" would be a good choice.

### LINKED ISSUE #1760: broken table border between playable and non-playable active games
On a virtualbox site which has master loaded:
broken_table_borders

Note the missing borders in the "Inactivity" column.

This is probably caused by the "Description" column.

### LINKED ISSUE #1761: Canceled games shouldn't show up in the Rejected table
If I create a game and cancel it, it shouldn't show up in the Rejected table.

(What should it do? I think ideally the "Successfully rejected game text at the top" should instead say "Successfully canceled Game N", with a link, so that if you clicked cancel by mistake, you could click the link, get the game page, and click something like "create a new game against <player> with the same parameters", much like the "create a new game" follow-up buttons after a completed game. So this would depend on the rejected-game page having some more stuff on it, but once it did, insta-dismissing but linking to that seems good.)

### PR DESCRIPTION
**Pull request by blackshadowshade**
Fixes #1747. Fixes #1758. Fixes #1760. Fixes #1761.

Depends on #1743.

This has been minimally tested for the ability to display rejected games and dismiss rejected games from the list.

It is intentional that rejecting a game does not automatically dismiss it as well.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/928/ (if it passes)

### CODE REVIEW

**Comment by blackshadowshade** [comment]
I'll update the pull for you once #1743 has been accepted, so that you get a clean diff.

**Comment by cgolubi1** [comment]
Drat.  This has merge conflicts.

**Comment by cgolubi1** [comment]
Hmm, looks like i can test this even though i can't merge it.  That's weird.

**Comment by cgolubi1** [comment]
Is there some reason the rejected games table doesn't include Max Wins even though the other tables do?  It feels distracting, different for the sake of being different.  Why not default to consistency?

many_tables

File: src/api/DummyApiResponder.php [hunk 1]
```diff
@@ -704,6 +723,48 @@ protected function get_interface_response_loadCompletedGames() {
...
+            'nTargetWinsArray' => array(),
+            'isAwaitingActionArray' => array(),
+            'gameStateArray' => array(),
+            'statusArray' => array(),
+            'inactivityArray' => array(),
+            'playerColorArray' => array(),
+            'opponentColorArray' => array(),
+        );
+
+        // game 5
```
Comment (cgolubi1) [inline src/api/DummyApiResponder.php]: Nit: you list this as game 505, so the comment should say game 505 too.  :>)

**Comment by cgolubi1** [comment]
Okay, so the UI draws a horizontal line above Active games... but this pull put rejected games below that horizontal line (and keeps New and Completed games above it).  That's obviously not right.  Can you move Rejected games above the line to go with New and Completed games?
wheres_that_hr

**Comment by dwvanstone** [comment]
In my opinion, if I'm the one who created a game it seems odd to have the option to "Reject" it in the New Games table. I think the better verb in this case is "Cancel" and it could immediately disappear. (It seems like an extra step for me to Reject it and then to have to Dismiss it.) Rejected Tables would then only be ones that somebody else declined the invitation.

**Comment by cgolubi1** [comment]
> In my opinion, if I'm the one who created a game it seems odd to have the option to "Reject" it in the New Games table. I think the better verb in this case is "Cancel" and it could immediately disappear. (It seems like an extra step for me to Reject it and then to have to Dismiss it.) Rejected Tables would then only be ones that somebody else declined the invitation.

I agree, but i also think that this is an edge case, not a prod-blocker, and the logical change to implement it could add more complexity and need more testing.  So my preference would be to make this a separate issue, and not block this pull on it.

**Comment by cgolubi1** [comment]
For anyone watching this: i can't merge this, and thus can't cleanly load it on dev, because of the merge conflict, but i was able to test it using vagrant, and i stood up a temporary public instance so other people could test it.  If you want to test this fix and i haven't e-mailed you, ping me, and i'll let you know where the temporary instance is.

**Comment by dwvanstone** [comment]
Just to be clear, what is the edge case you're talking about? When I create a game I will see a table that gives me the option to "Reject" it. That verb might be confusing.

**Comment by dwvanstone** [comment]
I think I know - the edge case is that if you do press it you will have to also dismiss it. (The other part I was compalining about.) I agree that is not the standard case.

**Comment by cgolubi1** [comment]
Yes, exactly, that's the edge case i meant.  For those playing the home game, i created #1754 for the "Reject should say Cancel" issue.

**Comment by cgolubi1** [comment]
To summarize the requests so far:
* The merge conflict needs to be fixed
* Nit: comment typo in DummyApiResponder.php
* The Rejected Games table is missing the "Max Wins" column --- bring the code in from the "New Games" table to generate that column for consistency
* The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it
* If you cancel a game you yourself created, it will show up in Rejected Games for you.  It should simply disappear from your view.  (We propose making a separate issue for this which is not a prod-blocker)

**Comment by irilyth** [comment]
Some comments!

> The Rejected Games table is missing the "Max Wins" column --- bring the code in from the "New Games" table to generate that column for consistency

An alternative idea: Why not make the columns the same in all the tables? Except for the final column, which is either Action, Completed, Rejected, or Inactivity, I think it'd be nice if all the others were the same across all of them. (So the New and Rejected tables should both have "Score // W/L/T (Max)" rather than just "Max wins"; among other things, it makes it easy to tell at a glance what that column is, since "0/0/0 (3)" is instantly recognizable in a way that just "3" isn't.

> The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it

Actually, I think it'd make more sense to have Rejected, Completed, the line, New, Active. That way, everything above the line is stuff that's in your court, but that no one but you cares about; and everything below the line is something that other people care about, with things that are in your court first (the second half of the Active table, with games that aren't in your court but that someone else cares about, is at the bottom, which seems right to me).

**Comment by cgolubi1** [comment]
> Actually, I think it'd make more sense to have Rejected, Completed, the line, New, Active

That sounds fine to me.

**Comment by cgolubi1** [comment]
Updating the summary, folding in irilyth's comments:

* The merge conflict needs to be fixed
* Nit: comment typo in DummyApiResponder.php
* The Rejected Games table is missing the "Max Wins" column --- all tables should have that column, and consider using the identical "Score (Max Wins)" syntax in all tables
* The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it --- possibly address this by making the table ordering "Rejected", "Completed", line, "New", "Active"
* If you cancel a game you yourself created, it will show up in Rejected Games for you. It should simply disappear from your view.  This has been split into #1761, and is not a blocker, but would be neat to address if it turned out to be trivial.

**Comment by blackshadowshade** [comment]
Okay, I've addressed all the points listed above. For the point about the horizontal line, I've removed it entirely for the moment. If we decide that we really need it, we'll need a new issue for that. However, I think that it adds visual complexity without actually having much benefit, especially since there are potentially five or six different tables present on the overview page at any time.

I've also rolled in fixes for #1761, #1760, and #1758, and I'll update the description at the top to add those fixes there.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/930/ (if it passes)

**Comment by cgolubi1** [comment]
In #1761, irilyth said: I think ideally the "Successfully rejected game text at the top" should instead say "Successfully canceled Game N", with a link, so that if you clicked cancel by mistake, you could click the link, get the game page, and click something like "create a new game against with the same parameters"...

Obviously, this fix doesn't do that (and can't until #1731 is addressed).  Is your expectation that he should open a new issue with that request?  Or should we modify this pull to note that it only partially addresses #1761.

**Comment by blackshadowshade** [comment]
Ah, I think I forgot to uncomment the expect(3) when I was getting the QUnit test to run. It does work.

As for the second part of #1761, I didn't notice that because it was in brackets.  :) I think that we should open a new issue with an explicitly different name (and possibly split it so that some of the content is added to #1731), so that it's easier to work out what is still to be done.

**Comment by cgolubi1** [comment]
Yes, i do still want the horizontal line (or some equivalent visual differentiator) --- opened #1763

**Comment by cgolubi1** [comment]
> The merge conflict needs to be fixed

Fixed

> Nit: comment typo in DummyApiResponder.php

Fixed

> The Rejected Games table is missing the "Max Wins" column --- all tables should have that column, and consider using the identical "Score (Max Wins)" syntax in all tables

The column now appears in all tables.  It is still "Max Wins" in the Rejected and New tables --- irilyth, if you still think it should be "Score (Max Wins)" in all tables, maybe open a separate issue for that?  IMO it's not a blocker, and i'd like to merge this pull if i can.

> The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it --- possibly address this by making the table ordering "Rejected", "Completed", line, "New", "Active"

Minimally done, opened #1763 for the rest

> If you cancel a game you yourself created, it will show up in Rejected Games for you. It should simply disappear from your view. This has been split into #1761, and is not a blocker, but would be neat to address if it turned out to be trivial.

Again, this was minimally done, and shadowshade requested that irilyth open a new issue with the parts of #1761 which were not yet done

**Comment by cgolubi1** [comment]
Jenkins for expect: http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/935/

**Comment by cgolubi1** [comment]
I'm not planning to wait for jenkins to finish --- i'll go ahead and merge, and start a pull to the dev site.  If it fails, we'll have another pull to fix the test.
